File: day01/day01.jl
#! /usr/local/bin/julia


function quantify(data; predicate = x -> x && true)
    mapreduce(predicate, +, data)
end


function p1()
    open("inputs/d01.txt", "r") do io
        data = map(s -> parse(Int64, s), eachline(io))
        return quantify(zip(data, data[2:end]), predicate = ((i, j),) -> j > i)
    end
end


function p2()
    open("inputs/d01.txt", "r") do io
        data = map(s -> parse(Int64, s), eachline(io))
        return quantify(
            zip(
                zip(data, data[2:end], data[3:end]),
                zip(data[2:end], data[3:end], data[4:end]),
            ),
            predicate = ((i, j),) -> sum(j) > sum(i),
        )
    end
end

println(p1())
println(p2())

File: day01/day01.py
#! /usr/bin/env python3

from helpers import *


data = process_inputs("01", int)

print(quantify(zip(data, data[1:]), lambda x: x[1] > x[0]))
print(
    quantify(
        zip(zip(data, data[1:], data[2:]), zip(data[1:], data[2:], data[3:])),
        lambda x: sum(x[1]) > sum(x[0]),
    )
)
